sql/catalog/lease: TestLeaseManagerLockedTimestampBasic failed

sql/catalog/lease.TestLeaseManagerLockedTimestampBasic [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20495911?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20495911?buildTab=artifacts#/) on master @ [3b1e9fa063036c453ff4ee103c633851186946cf](https://github.com/cockroachdb/cockroach/commits/3b1e9fa063036c453ff4ee103c633851186946cf):


```
created by github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).Init in goroutine 16504
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:54 +0x17c

goroutine 17438 [select]:
github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
	pkg/util/admission/work_queue.go:479 +0x140
created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 16504
	pkg/util/admission/work_queue.go:450 +0x5c

goroutine 16648 [chan receive, 14 minutes]:
github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).releaseLoop(0x4002ff1d60)
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:61 +0x6c
created by github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).Init in goroutine 16504
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:54 +0x17c

goroutine 16663 [chan receive, 14 minutes]:
github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).releaseLoop(0x40025a2fb0)
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:61 +0x6c
created by github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).Init in goroutine 16504
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:54 +0x17c

goroutine 16668 [select]:
github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).periodicGranter(0x4003c30fc0)
	external/com_github_cockroachdb_pebble/compaction_scheduler.go:414 +0x9c
created by github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).Register in goroutine 16504
	external/com_github_cockroachdb_pebble/compaction_scheduler.go:305 +0x98

goroutine 16666 [chan receive, 14 minutes]:
github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).releaseLoop(0x40024337b0)
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:61 +0x6c
created by github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).Init in goroutine 16504
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:54 +0x17c

goroutine 16649 [chan receive, 14 minutes]:
github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).releaseLoop(0x4002ff1d60)
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:61 +0x6c
created by github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).Init in goroutine 16504
	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:54 +0x17c

goroutine 16593 [select]:
github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).periodicGranter(0x40023864e0)
	external/com_github_cockroachdb_pebble/compaction_scheduler.go:414 +0x9c
created by github.com/cockroachdb/pebble.(*ConcurrencyLimitScheduler).Register in goroutine 16504
	external/com_github_cockroachdb_pebble/compaction_scheduler.go:305 +0x98

goroutine 17167 [select]:
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
	external/com_github_cockroachdb_pebble/vfs/disk_health.go:254 +0x98
created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 16504
	external/com_github_cockroachdb_pebble/vfs/disk_health.go:249 +0x64
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54615

sql/catalog/lease.TestLeaseManagerLockedTimestampBasic [failed](https://mesolite.cluster.engflow.com/invocations/default/ffe06f61-9670-40c1-9466-a00cf5c37e5f?testReportRun=1&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvbGVhc2U6bGVhc2VfdGVzdA==) on master @ [223617d5f65835071f1f3395bc7b487e6dc029c9](https://github.com/cockroachdb/cockroach/commits/223617d5f65835071f1f3395bc7b487e6dc029c9):


```
=== RUN   TestLeaseManagerLockedTimestampBasic
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestLeaseManagerLockedTimestampBasic1297632801
    test_log_scope.go:82: use -show-logs to present logs inline
    lease_internal_test.go:1806: 
        	Error Trace:	pkg/sql/catalog/lease/lease_internal_test.go:1806
        	Error:      	Not equal: 
        	            	expected: [][]string{[]string{"t1", "COLUMN", "DELETE_ONLY"}}
        	            	actual  : [][]string{[]string{"t1", "COLUMN", "WRITE_ONLY"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	   (string) (len=6) "COLUMN",
        	            	-  (string) (len=11) "DELETE_ONLY"
        	            	+  (string) (len=10) "WRITE_ONLY"
        	            	  }
        	Test:       	TestLeaseManagerLockedTimestampBasic
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseManagerLockedTimestampBasic1297632801
--- FAIL: TestLeaseManagerLockedTimestampBasic (3.24s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/catalog/lease.TestLeaseManagerLockedTimestampBasic [failed](https://mesolite.cluster.engflow.com/invocations/default/20d6ca98-78cf-4dcd-961d-1a744600a6dc?testReportRun=2&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvbGVhc2U6bGVhc2VfdGVzdA==) on master @ [223617d5f65835071f1f3395bc7b487e6dc029c9](https://github.com/cockroachdb/cockroach/commits/223617d5f65835071f1f3395bc7b487e6dc029c9):


```
=== RUN   TestLeaseManagerLockedTimestampBasic
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestLeaseManagerLockedTimestampBasic2263452198
    test_log_scope.go:82: use -show-logs to present logs inline
    lease_internal_test.go:1806: 
        	Error Trace:	pkg/sql/catalog/lease/lease_internal_test.go:1806
        	Error:      	Not equal: 
        	            	expected: [][]string{[]string{"t1", "COLUMN", "DELETE_ONLY"}}
        	            	actual  : [][]string{[]string{"t1", "COLUMN", "WRITE_ONLY"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	   (string) (len=6) "COLUMN",
        	            	-  (string) (len=11) "DELETE_ONLY"
        	            	+  (string) (len=10) "WRITE_ONLY"
        	            	  }
        	Test:       	TestLeaseManagerLockedTimestampBasic
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseManagerLockedTimestampBasic2263452198
--- FAIL: TestLeaseManagerLockedTimestampBasic (3.29s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/catalog/lease.TestLeaseManagerLockedTimestampBasic [failed](https://mesolite.cluster.engflow.com/invocations/default/911077b6-93c7-4aa8-b28e-46bfda2cf53c?testReportRun=2&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvbGVhc2U6bGVhc2VfdGVzdA==) on master @ [3476899d48145434a9ea53f95c537e670927bd26](https://github.com/cockroachdb/cockroach/commits/3476899d48145434a9ea53f95c537e670927bd26):


```
=== RUN   TestLeaseManagerLockedTimestampBasic
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestLeaseManagerLockedTimestampBasic253433344
    test_log_scope.go:82: use -show-logs to present logs inline
    lease_internal_test.go:1806: 
        	Error Trace:	pkg/sql/catalog/lease/lease_internal_test.go:1806
        	Error:      	Not equal: 
        	            	expected: [][]string{[]string{"t1", "COLUMN", "DELETE_ONLY"}}
        	            	actual  : [][]string{[]string{"t1", "COLUMN", "WRITE_ONLY"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	   (string) (len=6) "COLUMN",
        	            	-  (string) (len=11) "DELETE_ONLY"
        	            	+  (string) (len=10) "WRITE_ONLY"
        	            	  }
        	Test:       	TestLeaseManagerLockedTimestampBasic
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseManagerLockedTimestampBasic253433344
--- FAIL: TestLeaseManagerLockedTimestampBasic (3.39s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/backup/backup_test.go
+++ b/pkg/backup/backup_test.go
@@ -6687,7 +6687,6 @@ INSERT INTO baz.bar VALUES (110, 'a'), (210, 'b'), (310, 'c'), (410, 'd'), (510,
 	tenant10.Exec(t, `BACKUP DATABASE baz INTO 'userfile://defaultdb.myfililes/test4' with revision_history`)
 	expected = nil
 	for _, resume := range []exportResumePoint{
-		{mkSpan(id2, "/Tenant/10/Table/3", "/Tenant/10/Table/4"), withoutTS},
 		{mkSpan(id2, "/Tenant/10/Table/:id/1", "/Tenant/10/Table/:id/2"), withoutTS},
 		{mkSpan(id2, "/Tenant/10/Table/:id/1/210", "/Tenant/10/Table/:id/2"), withoutTS},
 		// We have two entries for 210 because of history and super small table

--- a/pkg/backup/targets.go
+++ b/pkg/backup/targets.go
@@ -185,7 +185,7 @@ func getAllDescChanges(
 	startTime, endTime hlc.Timestamp,
 	priorIDs map[descpb.ID]descpb.ID,
 ) ([]backuppb.BackupManifest_DescriptorRevision, error) {
-	startKey := codec.TablePrefix(keys.DescriptorTableID)
+	startKey := codec.IndexPrefix(keys.DescriptorTableID, keys.DescriptorTablePrimaryKeyIndexID)
 	endKey := startKey.PrefixEnd()
 
 	g := ctxgroup.WithContext(ctx)

--- a/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
+++ b/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
@@ -831,7 +831,7 @@ func (tf *schemaFeed) fetchDescriptorVersions(
 	}
 	codec := tf.leaseMgr.Codec()
 	start := crtime.NowMono()
-	span := roachpb.Span{Key: codec.TablePrefix(keys.DescriptorTableID)}
+	span := roachpb.Span{Key: codec.IndexPrefix(keys.DescriptorTableID, keys.DescriptorTablePrimaryKeyIndexID)}
 	span.EndKey = span.Key.PrefixEnd()
 
 	tf.mu.Lock()

--- a/pkg/ccl/changefeedccl/schemafeed/schematestutils/schema_test_utils.go
+++ b/pkg/ccl/changefeedccl/schemafeed/schematestutils/schema_test_utils.go
@@ -148,7 +148,7 @@ func FetchDescVersionModificationTime(
 	db := serverutils.OpenDBConn(
 		t, s.SQLAddr(), dbName, false, s.AppStopper())
 
-	tblKey := s.Codec().TablePrefix(keys.DescriptorTableID)
+	tblKey := s.Codec().IndexPrefix(keys.DescriptorTableID, keys.DescriptorTablePrimaryKeyIndexID)
 	header := kvpb.RequestHeader{
 		Key:    tblKey,
 		EndKey: tblKey.PrefixEnd(),

--- a/pkg/config/system.go
+++ b/pkg/config/system.go
@@ -211,9 +211,9 @@ func (s *SystemConfig) GetLargestObjectID(
 	// Search for the descriptor table entries within the SystemConfig. lowIndex
 	// (in s.Values) is the first and highIndex one past the last KV pair in the
 	// descriptor table.
-	lowBound := keys.SystemSQLCodec.TablePrefix(keys.DescriptorTableID)
+	lowBound := keys.SystemSQLCodec.IndexPrefix(keys.DescriptorTableID, keys.DescriptorTablePrimaryKeyIndexID)
 	lowIndex := s.getIndexBound(lowBound)
-	highBound := keys.SystemSQLCodec.TablePrefix(keys.DescriptorTableID + 1)
+	highBound := keys.SystemSQLCodec.IndexPrefix(keys.DescriptorTableID+1, keys.DescriptorTablePrimaryKeyIndexID)
 	highIndex := s.getIndexBound(highBound)
 	if lowIndex == highIndex {
 		return 0, fmt.Errorf("descriptor table not found in system config of %d values", len(s.Values))

--- a/pkg/keys/constants.go
+++ b/pkg/keys/constants.go
@@ -426,9 +426,14 @@ const (
 	ZonesTableConfigColumnID = 2
 	ZonesTableConfigColFamID = 2
 
-	DescriptorTablePrimaryKeyIndexID         = 1
-	DescriptorTableDescriptorColID           = 2
-	DescriptorTableDescriptorColFamID        = 2
+	DescriptorTablePrimaryKeyIndexID  = 1
+	DescriptorTableDescriptorColID    = 2
+	DescriptorTableDescriptorColFamID = 2
+	// DescriptorTableDescriptorUpdateIndexID is not a real index. It is a special
+	// ID used to construct index entries that inform the lease subsystem of
+	// descriptor updates within a transaction. The value for such an entry is a
+	// descpb.DescriptorUpdates message.
+	DescriptorTableDescriptorUpdateIndexID   = 2
 	TenantsTablePrimaryKeyIndexID            = 1
 	SpanConfigurationsTablePrimaryKeyIndexID = 1
 	CommentsTablePrimaryKeyIndexID           = 1

--- a/pkg/keys/sql.go
+++ b/pkg/keys/sql.go
@@ -209,12 +209,26 @@ func (e sqlEncoder) DescMetadataPrefix() roachpb.Key {
 	return e.IndexPrefix(DescriptorTableID, DescriptorTablePrimaryKeyIndexID)
 }
 
+// DescUpdatePrefix returns the key prefix for all descriptors in the
+// system.descriptor table.
+func (e sqlEncoder) DescUpdatePrefix() roachpb.Key {
+	return e.IndexPrefix(DescriptorTableID, DescriptorTableDescriptorUpdateIndexID)
+}
+
 // DescMetadataKey returns the key for the descriptor in the system.descriptor
 // table.
 func (e sqlEncoder) DescMetadataKey(descID uint32) roachpb.Key {
 	k := e.DescMetadataPrefix()
 	k = encoding.EncodeUvarintAscending(k, uint64(descID))
 	return MakeFamilyKey(k, DescriptorTableDescriptorColFamID)
+}
+
+// DescMetadataUpdateKey returns the key for the descriptor in the system.descriptor
+// table.
+func (e sqlEncoder) DescMetadataUpdateKey(descID uint32) roachpb.Key {
+	k := e.DescUpdatePrefix()
+	k = encoding.EncodeUvarintAscending(k, uint64(descID))
+	return MakeFamilyKey(k, 0)
 }
 
 // TenantMetadataKey returns the key for the tenant metadata in the
@@ -305,6 +319,19 @@ func (d sqlDecoder) DecodeDescMetadataID(key roachpb.Key) (uint32, error) {
 		return 0, errors.Errorf("descriptor ID %d exceeds uint32 bounds", id)
 	}
 	return uint32(id), nil
+}
+
+// DecodeDescUpdateKey decodes a descriptor ID from a descriptor metadata key.
+func (d sqlDecoder) DecodeDescUpdateKey(key roachpb.Key) (bool, error) {
+	// Extract table and index ID from key.
+	_, tableID, indexID, err := d.DecodeIndexPrefix(key)
+	if err != nil {
+		return false, err
+	}
+	if tableID != DescriptorTableID {
+		return false, errors.Errorf("key is not a descriptor table entry: %v", key)
+	}
+	return indexID == DescriptorTableDescriptorUpdateIndexID, nil
 }
 
 // DecodeTenantMetadataID decodes a tenant ID from a tenant metadata key.

--- a/pkg/server/diagnostics/reporter.go
+++ b/pkg/server/diagnostics/reporter.go
@@ -449,7 +449,7 @@ func (r *Reporter) populateSQLInfo(uptime int64, sql *diagnosticspb.SQLInstanceI
 // type fields. Check out `schematelemetry` package for a better data source for
 // collecting redacted schema information.
 func (r *Reporter) collectSchemaInfo(ctx context.Context) ([]descpb.TableDescriptor, error) {
-	startKey := keys.MakeSQLCodec(r.TenantID).TablePrefix(keys.DescriptorTableID)
+	startKey := keys.MakeSQLCodec(r.TenantID).IndexPrefix(keys.DescriptorTableID, keys.DescriptorTablePrimaryKeyIndexID)
 	endKey := startKey.PrefixEnd()
 	kvs, err := r.DB.Scan(ctx, startKey, endKey, 0)
 	if err != nil {

--- a/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go
+++ b/pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go
@@ -195,6 +195,15 @@ func (s *SQLWatcher) watchForDescriptorUpdates(
 			// Event for a tombstone on a tombstone -- nothing for us to do here.
 			return
 		}
+		// Skip over any modifications to the descriptor update tracking key, this
+		// is transaction information for the lease manager only.
+		if isUpdateKey, err := s.codec.DecodeDescUpdateKey(ev.Key); isUpdateKey || err != nil {
+			if err != nil {
+				log.Dev.Warningf(ctx, "failed to decode descriptor update key: %v", err)
+			}
+			return
+		}
+
 		value := ev.Value
 		if !ev.Value.IsPresent() {
 			// The descriptor was deleted.

--- a/pkg/sql/catalog/catalogkeys/keys.go
+++ b/pkg/sql/catalog/catalogkeys/keys.go
@@ -287,6 +287,11 @@ func MakeDescMetadataKey(codec keys.SQLCodec, descID descpb.ID) roachpb.Key {
 	return codec.DescMetadataKey(uint32(descID))
 }
 
+// MakeDescUpdateKey returns the key for the descriptor.
+func MakeDescUpdateKey(codec keys.SQLCodec, id descpb.ID) roachpb.Key {
+	return codec.DescMetadataUpdateKey(uint32(id))
+}
+
 // CommentsMetadataPrefix returns the key prefix for all comments in the
 // system.comments table.
 func CommentsMetadataPrefix(codec keys.SQLCodec) roachpb.Key {

--- a/pkg/sql/catalog/descpb/structured.proto
+++ b/pkg/sql/catalog/descpb/structured.proto
@@ -1937,3 +1937,14 @@ message Descriptor {
     FunctionDescriptor function = 5;
   }
 }
+
+// DescriptorUpdates tracks updates to descriptors within a transaction, which
+// will be written to a special key.
+message DescriptorUpdates {
+  option (gogoproto.equal) = true;
+  // Needed for the descriptorProto interface.
+  option (gogoproto.goproto_getters) = true;
+
+  repeated uint32 descriptor_ids = 1 [(gogoproto.customname) = "DescriptorIDs", (gogoproto.casttype) = "ID"];
+  repeated uint32 descriptor_versions = 2  [(gogoproto.customname) = "DescriptorVersions", (gogoproto.casttype) = "DescriptorVersion"];
+}

--- a/pkg/sql/catalog/descs/collection.go
+++ b/pkg/sql/catalog/descs/collection.go
@@ -436,6 +436,36 @@ func (tc *Collection) maybeMarkVersionBumpOnly(
 			isVersionBumpOnly
 }
 
+// EmitDescriptorUpdatesKey writes a special key that helps the lease manager
+// know which descriptors were updated by this transaction. This allows it to
+// transactionally make these descriptors available.
+func (tc *Collection) EmitDescriptorUpdatesKey(ctx context.Context, txn *kv.Txn) error {
+	// This key is only emitted for 25.4 and above. Additionally, PCR
+	// reader catalog can update a large number of descriptors in a transaction,
+	// and are exempt from this logic.
+	if !tc.settings.Version.IsActive(ctx, clusterversion.V25_4) ||
+		tc.readerCatalogSetup ||
+		tc.uncommitted.uncommitted.Len() == 0 ||
+		!lease.LockedLeaseTimestamp.Get(&tc.settings.SV) {
+		return nil
+	}
+	updates := &descpb.DescriptorUpdates{}
+	descUpdateID := descpb.InvalidID
+	// Add all the descriptors that have been modified in this transaction.
+	if err := tc.uncommitted.iterateUncommittedByID(func(desc catalog.Descriptor) error {
+		updates.DescriptorIDs = append(updates.DescriptorIDs, desc.GetID())
+		updates.DescriptorVersions = append(updates.DescriptorVersions, desc.GetVersion())
+		if descUpdateID < desc.GetID() {
+			descUpdateID = desc.GetID()
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	descUpdateKey := catalogkeys.MakeDescUpdateKey(tc.codec(), descUpdateID)
+	return txn.Put(ctx, descUpdateKey, updates)
+}
+
 // WriteDescToBatch calls MaybeIncrementVersion, adds the descriptor to the
 // collection as an uncommitted descriptor, and writes it into b.
 func (tc *Collection) WriteDescToBatch(

--- a/pkg/sql/catalog/lease/BUILD.bazel
+++ b/pkg/sql/catalog/lease/BUILD.bazel
@@ -69,6 +69,7 @@ go_library(
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_cockroachdb_logtags//:logtags",
         "@com_github_cockroachdb_redact//:redact",
+        "@com_github_google_btree//:btree",
     ],
 )
 

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -59,6 +59,7 @@ import (
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/logtags"
 	"github.com/cockroachdb/redact"
+	"github.com/google/btree"
 )
 
 var errRenewLease = errors.New("renew lease on id")
@@ -684,8 +685,13 @@ func ensureVersion(
 		return err
 	}
 
-	if s := m.findNewest(id); s != nil && s.GetVersion() < minVersion {
+	s := m.findNewest(id)
+	if s != nil && s.GetVersion() < minVersion {
 		return errors.Errorf("version %d for descriptor %s does not exist yet", minVersion, s.GetName())
+	} else if s != nil {
+		// Process any descriptor updates that will allow us to move the close timestamp
+		// forward.
+		m.processDescriptorUpdate(ctx, id, s.GetVersion(), s.GetModificationTime())
 	}
 	return nil
 }
@@ -1327,6 +1333,20 @@ const (
 	AcquireBackground
 )
 
+// descriptorTxnUpdate tracks updates to a set of descriptors in the
+// system.descriptors by a transaction, that should be made available
+// atomically.
+type descriptorTxnUpdate struct {
+	mu struct {
+		syncutil.Mutex
+		// DescriptorUpdates contains any descriptor versions that we are
+		// presently waiting for.
+		descpb.DescriptorUpdates
+	}
+	timestamp hlc.Timestamp
+	key       roachpb.Key
+}
+
 // Manager manages acquiring and releasing per-descriptor leases. It also
 // handles resolving descriptor names to descriptor IDs. The leases are managed
 // internally with a descriptor and expiration time exported by the
@@ -1365,6 +1385,14 @@ type Manager struct {
 
 		// rangeFeedRestartInProgress tracks if a range feed restart is in progress.
 		rangeFeedRestartInProgress bool
+
+		// descriptorTxnUpdatesToProcess is a list of descriptor txn updates that
+		// are waiting for new descriptor versions to be available.
+		descriptorTxnUpdatesToProcess *btree.BTreeG[*descriptorTxnUpdate]
+
+		// hasUpdatesToDelete tracks if there is data in the key space that
+		// should be cleaned up.
+		hasUpdatesToDelete bool
 	}
 
 	// closeTimeStamp for the range feed, which is the timestamp
@@ -1387,6 +1415,9 @@ type Manager struct {
 	descUpdateCh chan catalog.Descriptor
 	// descDelCh receives deleted descriptors from the range feed.
 	descDelCh chan descpb.ID
+	// descMetaDataUpdateCh receives updated transaction metadata from the
+	// range feed.
+	descMetaDataUpdateCh chan *descriptorTxnUpdate
 	// rangefeedErrCh receives any terminal errors from the rangefeed.
 	rangefeedErrCh chan error
 	// leaseGeneration increments any time a new or existing descriptor is
@@ -1523,6 +1554,10 @@ func NewLeaseManager(
 	lm.stopper.AddCloser(lm.sem.Closer("stopper"))
 	lm.stopper.AddCloser(stop.CloserFn(lm.AssertAllLeasesAreReleasedAfterDrain))
 	lm.mu.descriptors = make(map[descpb.ID]*descriptorState)
+	lm.mu.descriptorTxnUpdatesToProcess = btree.NewG(2, func(a, b *descriptorTxnUpdate) bool {
+		return a.timestamp.Less(b.timestamp)
+	})
+	lm.mu.hasUpdatesToDelete = true
 	lm.waitForInit = make(chan struct{})
 	// We are going to start the range feed later when StartRefreshLeasesTask
 	// is invoked inside pre-start. So, that guarantees all range feed events
@@ -1533,6 +1568,7 @@ func NewLeaseManager(
 	lm.draining.Store(false)
 	lm.descUpdateCh = make(chan catalog.Descriptor)
 	lm.descDelCh = make(chan descpb.ID)
+	lm.descMetaDataUpdateCh = make(chan *descriptorTxnUpdate)
 	lm.rangefeedErrCh = make(chan error)
 	lm.bytesMonitor = mon.NewMonitor(mon.Options{
 		Name:       mon.MakeName("leased-descriptors"),
@@ -1948,6 +1984,131 @@ func (m *Manager) findDescriptorState(id descpb.ID, create bool) *descriptorStat
 	return t
 }
 
+// hasDescUpdateBeenApplied checks if all versions of a descriptor update are
+// present. If not, it returns a new descpb.DescriptorUpdates containing the
+// versions that are still missing.
+func (m *Manager) hasDescUpdateBeenApplied(
+	updates descpb.DescriptorUpdates,
+) (allApplied bool, remaining descpb.DescriptorUpdates) {
+	allApplied = true
+	for idx, descID := range updates.DescriptorIDs {
+		descVersionState := m.findNewest(descID)
+		requiredVersion := updates.DescriptorVersions[idx]
+		// If a descriptor is missing, we still consider it as applied, since
+		// we will lease a newer version.
+		if descVersionState != nil && descVersionState.GetVersion() < requiredVersion {
+			allApplied = allApplied && descVersionState == nil
+			remaining.DescriptorIDs = append(remaining.DescriptorIDs, descID)
+			remaining.DescriptorVersions = append(remaining.DescriptorVersions, requiredVersion)
+		}
+	}
+	return allApplied, remaining
+}
+
+// getDescriptorTxnUpdateEntry returns the descriptorTxnUpdate entry for the given timestamp,
+// if there are any pending updates.
+func (m *Manager) getDescriptorTxnUpdateEntry(timestamp hlc.Timestamp) *descriptorTxnUpdate {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	key := &descriptorTxnUpdate{timestamp: timestamp}
+	entry, hasTS := m.mu.descriptorTxnUpdatesToProcess.Get(key)
+	if !hasTS {
+		return nil
+	}
+	return entry
+}
+
+// processDescriptorUpdate will process this descriptor update for txn
+// consistency.
+func (m *Manager) processDescriptorUpdate(
+	ctx context.Context, id descpb.ID, version descpb.DescriptorVersion, timestamp hlc.Timestamp,
+) {
+	entry := m.getDescriptorTxnUpdateEntry(timestamp)
+	// This timestamp has no pending update.
+	if entry == nil {
+		return
+	}
+	noEntriesLeft := func() bool {
+		// Lock this specific entry
+		entry.mu.Lock()
+		defer entry.mu.Unlock()
+		// Process this descriptor version inside the entry.
+		for idx, descID := range entry.mu.DescriptorIDs {
+			if descID != id {
+				continue
+			}
+			// The target version has not arrived yet for this update.
+			if entry.mu.DescriptorVersions[idx] > version {
+				return false
+			}
+			// Otherwise, let's move this version to the end first.
+			endIdx := len(entry.mu.DescriptorIDs) - 1
+			if idx != endIdx {
+				entry.mu.DescriptorIDs[idx], entry.mu.DescriptorIDs[endIdx] =
+					entry.mu.DescriptorIDs[endIdx], entry.mu.DescriptorIDs[idx]
+				entry.mu.DescriptorVersions[idx], entry.mu.DescriptorVersions[endIdx] =
+					entry.mu.DescriptorVersions[endIdx], entry.mu.DescriptorVersions[idx]
+			}
+			// Remove one element from the end after.
+			entry.mu.DescriptorIDs = entry.mu.DescriptorIDs[:len(entry.mu.DescriptorIDs)-1]
+			entry.mu.DescriptorVersions = entry.mu.DescriptorVersions[:len(entry.mu.DescriptorIDs)-1]
+			break
+		}
+		return len(entry.mu.DescriptorIDs) == 0
+	}()
+	// Everything is processed, so we can move the timestamp forward.
+	if noEntriesLeft {
+		// If the length is zero by this point, then we can move the timestamp forward
+		// and remove this entry.
+		targetTimestamp := m.markDescriptorUpdatesAsComplete()
+		if !targetTimestamp.IsEmpty() {
+			m.advanceCloseTimestamp(targetTimestamp)
+		}
+	}
+}
+
+// markDescriptorUpdatesAsComplete will move the close timestamp forward if
+// all descriptor updates have been applied.
+func (m *Manager) markDescriptorUpdatesAsComplete() hlc.Timestamp {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	timestamp := hlc.Timestamp{}
+	for {
+		minUpdate, hasMin := m.mu.descriptorTxnUpdatesToProcess.Min()
+		if !hasMin {
+			break
+		}
+		shouldRemoveEntry := func() bool {
+			minUpdate.mu.Lock()
+			defer minUpdate.mu.Unlock()
+			return len(minUpdate.mu.DescriptorIDs) == 0
+		}()
+		if !shouldRemoveEntry {
+			break
+		}
+		timestamp = minUpdate.timestamp
+		m.mu.descriptorTxnUpdatesToProcess.Delete(minUpdate)
+	}
+	return timestamp
+}
+
+// advanceCloseTimestamp advances the close timestamp atomically, if the current
+// close timestamp is smaller.
+func (m *Manager) advanceCloseTimestamp(timestamp hlc.Timestamp) {
+	for {
+		oldTimestamp := m.closeTimestamp.Load().(hlc.Timestamp)
+		// Timestamp has already been advanced.
+		if !oldTimestamp.Less(timestamp) {
+			return
+		}
+		// Otherwise, attempt to swap the timestamp, if successful we
+		// will return.
+		if m.closeTimestamp.CompareAndSwap(oldTimestamp, timestamp) {
+			return
+		}
+	}
+}
+
 // StartRefreshLeasesTask starts a goroutine that refreshes the lease manager
 // leases for descriptors received in the latest system configuration via gossip or
 // rangefeeds. This function must be passed a non-nil gossip if
@@ -2062,6 +2223,30 @@ func (m *Manager) StartRefreshLeasesTask(ctx context.Context, s *stop.Stopper, d
 				if evFunc := m.testingKnobs.TestingDescriptorRefreshedEvent; evFunc != nil {
 					evFunc(desc.DescriptorProto())
 				}
+			case descUpdate := <-m.descMetaDataUpdateCh:
+				// Check if the update has been applied.
+				allApplied, remaining := m.hasDescUpdateBeenApplied(descUpdate.mu.DescriptorUpdates)
+				descUpdate.mu.DescriptorUpdates = remaining
+				advanceTimestamp := func() bool {
+					m.mu.Lock()
+					defer m.mu.Unlock()
+					m.mu.hasUpdatesToDelete = true
+					// If there are no other earlier pending updates, advance the timestamp.
+					minUpdate, hasMin := m.mu.descriptorTxnUpdatesToProcess.Min()
+					if allApplied && (!hasMin ||
+						!minUpdate.timestamp.Less(descUpdate.timestamp)) {
+						return true
+					}
+					// Otherwise, insert this into the queue of pending updates
+					m.mu.descriptorTxnUpdatesToProcess.ReplaceOrInsert(descUpdate)
+					return false
+				}()
+				if advanceTimestamp {
+					if m.testingKnobs.TestingOnUpdateReadTimestamp != nil {
+						m.testingKnobs.TestingOnUpdateReadTimestamp(descUpdate.timestamp)
+					}
+					m.advanceCloseTimestamp(descUpdate.timestamp)
+				}
 			case <-s.ShouldQuiesce():
 				return
 			}
@@ -2113,13 +2298,43 @@ func (m *Manager) watchForUpdates(ctx context.Context) {
 		if m.testingKnobs.DisableRangeFeedCheckpoint {
 			return
 		}
-		if len(ev.Value.RawBytes) == 0 {
+		// Check first if it is the special descriptor metadata key used to track
+		// descriptors modified by transactions.
+		if isUpdateKey, err := m.Codec().DecodeDescUpdateKey(ev.Key); err != nil || isUpdateKey {
+			if err != nil {
+				log.Dev.Warningf(ctx, "unable to decode update metadata key %v", ev.Key)
+				return
+			}
+			// Ignore deletes on this key space.
+			if len(ev.Value.RawBytes) == 0 {
+				return
+			}
+			// Otherwise, this is a descriptor update key.
+			var descUpdates descpb.DescriptorUpdates
+			err := ev.Value.GetProto(&descUpdates)
+			if err != nil {
+				log.Dev.Warningf(ctx, "unable to decode descriptor update value %v", err)
+				return
+			}
+			update := &descriptorTxnUpdate{
+				key:       ev.Key,
+				timestamp: ev.Timestamp(),
+			}
+			update.mu.DescriptorUpdates = descUpdates
+			select {
+			case <-m.stopper.ShouldQuiesce():
+			case <-ctx.Done():
+			case m.descMetaDataUpdateCh <- update:
+			}
+			return
+		} else if len(ev.Value.RawBytes) == 0 {
 			id, err := m.Codec().DecodeDescMetadataID(ev.Key)
 			if err != nil {
 				log.Dev.Infof(ctx, "unable to decode metadata key %v", ev.Key)
 				return
 			}
 			select {
+			case <-m.stopper.ShouldQuiesce():
 			case <-ctx.Done():
 			case m.descDelCh <- descpb.ID(id):
 			}
@@ -2139,14 +2354,15 @@ func (m *Manager) watchForUpdates(ctx context.Context) {
 				ev.Key, mut.GetID(), mut.GetName(), mut.GetVersion())
 		}
 		select {
+		case <-m.stopper.ShouldQuiesce():
 		case <-ctx.Done():
 		case m.descUpdateCh <- mut:
 		}
 	}
 
 	handleCheckpoint := func(ctx context.Context, checkpoint *kvpb.RangeFeedCheckpoint) {
-		if m.testingKnobs.TestingOnRangeFeedCheckPoint != nil {
-			m.testingKnobs.TestingOnRangeFeedCheckPoint()
+		if m.testingKnobs.TestingOnUpdateReadTimestamp != nil {
+			m.testingKnobs.TestingOnUpdateReadTimestamp(checkpoint.ResolvedTS)
 		}
 		// Track checkpoints that occur from the rangefeed to make sure progress
 		// is always made.
@@ -2156,7 +2372,15 @@ func (m *Manager) watchForUpdates(ctx context.Context) {
 			return
 		}
 		m.mu.rangeFeedCheckpoints += 1
-		m.closeTimestamp.Store(checkpoint.ResolvedTS)
+		m.advanceCloseTimestamp(checkpoint.ResolvedTS)
+		// Clean up all entries before the resolve timestamp.
+		for {
+			minTS, hasMin := m.mu.descriptorTxnUpdatesToProcess.Min()
+			if !hasMin || !minTS.timestamp.Less(checkpoint.ResolvedTS) {
+				break
+			}
+			m.mu.descriptorTxnUpdatesToProcess.Delete(minTS)
+		}
 	}
 
 	// Assert that the range feed is already terminated.
@@ -2257,6 +2481,10 @@ func (m *Manager) RunBackgroundLeasingTask(ctx context.Context) {
 		rangeFeedProgressWatchDogTimeout,
 			rangeFeedProgressWatchDogEnabled := m.getRangeFeedMonitorSettings()
 		rangeFeedProgressWatchDog.Reset(rangeFeedProgressWatchDogTimeout)
+		// Descriptor update clean-up timer
+		var descriptorUpdateCleanupTimer timeutil.Timer
+		descriptorUpdateCleanupTimerDuration := 30 * time.Minute
+		descriptorUpdateCleanupTimer.Reset(descriptorUpdateCleanupTimerDuration)
 		for {
 			select {
 			case <-m.stopper.ShouldQuiesce():
@@ -2289,6 +2517,10 @@ func (m *Manager) RunBackgroundLeasingTask(ctx context.Context) {
 
 				// Clean up session based leases that have expired.
 				m.cleanupExpiredSessionLeases(ctx)
+			case <-descriptorUpdateCleanupTimer.C:
+				// Clean up the update key space.
+				m.cleanupUpdateKeys(ctx)
+				descriptorUpdateCleanupTimer.Reset(descriptorUpdateCleanupTimerDuration)
 			}
 		}
 	})
@@ -2396,6 +2628,27 @@ func (m *Manager) cleanupExpiredSessionLeases(ctx context.Context) {
 		}); err != nil {
 			log.Dev.Infof(ctx, "unable to delete leases from storage %s", err)
 		}
+	}
+}
+
+// cleanupUpdateKeys updates special keys that exist for tracking descriptor
+// updates within transactions.
+func (m *Manager) cleanupUpdateKeys(ctx context.Context) {
+	// Only clean update keys if we received any range feed updates.
+	m.mu.Lock()
+	hasUpdates := m.mu.hasUpdatesToDelete
+	m.mu.Unlock()
+	if !hasUpdates {
+		return
+	}
+	// Issue a DelRange on this key space.
+	err := m.storage.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		prefix := m.storage.codec.DescUpdatePrefix()
+		_, err := txn.KV().DelRange(ctx, prefix, prefix.PrefixEnd(), false)
+		return err
+	})
+	if err != nil {
+		log.Dev.Infof(ctx, "unable to delete descriptor update keys from storage: %v", err)
 	}
 }
 
@@ -2514,6 +2767,7 @@ func (m *Manager) DeleteOrphanedLeases(
 		retryOpts.MaxRetries = 10
 		m.deleteOrphanedLeasesFromStaleSession(ctx, retryOpts, timeThreshold, locality)
 		m.deleteOrphanedLeasesWithSameInstanceID(ctx, retryOpts, timeThreshold, instanceID)
+		m.cleanupUpdateKeys(ctx)
 	})
 }
 

--- a/pkg/sql/catalog/lease/lease_internal_test.go
+++ b/pkg/sql/catalog/lease/lease_internal_test.go
@@ -30,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/sqllivenesstestutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
@@ -1745,6 +1746,8 @@ func TestLeaseManagerLockedTimestampBasic(t *testing.T) {
 
 	var blockUpdates atomic.Bool
 	var blockCheckPoint atomic.Bool
+	var targetTS atomic.Value
+	targetTS.Store(hlc.Timestamp{})
 	updateCh := make(chan struct{})
 	rangeFeedCh := make(chan struct{})
 
@@ -1763,17 +1766,22 @@ func TestLeaseManagerLockedTimestampBasic(t *testing.T) {
 							return
 						}
 						<-updateCh
+						//nolint:descriptormarshal
+						targetTS.Store(descriptor.GetTable().GetModificationTime())
 						blockCheckPoint.Store(true)
 					},
-					TestingOnRangeFeedCheckPoint: func() {
+					TestingOnUpdateReadTimestamp: func(timestamp hlc.Timestamp) {
 						if !blockCheckPoint.Load() {
 							return
 						}
-						blockCheckPoint.Store(false)
+						ts := targetTS.Load().(hlc.Timestamp)
+						if ts.IsEmpty() || timestamp.Less(ts) {
+							return
+						}
 						<-rangeFeedCh
+						targetTS.Store(hlc.Timestamp{})
 					},
 				},
-				// FIXME: Add checkpoint function.
 			},
 			Settings: st,
 		})
@@ -1785,36 +1793,62 @@ func TestLeaseManagerLockedTimestampBasic(t *testing.T) {
 	r.Exec(t, "INSERT INTO t1 VALUES (1)")
 
 	grp := ctxgroup.WithContext(context.Background())
+	blockUpdates.Store(true)
 	grp.GoCtx(func(ctx context.Context) error {
 		_, err := db.Exec("ALTER TABLE t1 ADD COLUMN n2 int")
 		return err
 	})
 	go func() {
 	}()
+	var id int
+	r.QueryRow(t, "SELECT 't1'::REGCLASS::OID;").Scan(&id)
+	lm := srv.LeaseManager().(*Manager)
 
-	//	var id int
-	//	r.QueryRow(t, "SELECT 'd1.public.t1'::REGCLASS::OID;").Scan(&id)
+	// Note: We only need to hold old leases because the logic for doing this
+	// automatically is not merged yet. Once it is merged, the release of them
+	// will be delayed until the timestamp is updated.
+	var heldDescriptors []LeasedDescriptor
+	releaseHeldDescriptors := func() {
+		for _, ld := range heldDescriptors {
+			ld.Release(ctx)
+		}
+		heldDescriptors = nil
+	}
+	defer releaseHeldDescriptors()
 
-	// FIXME: Add
-	// 1) SELECT , confirm version, allow check point, confirm version...3
+	getDescriptorVersion := func() descpb.DescriptorVersion {
+		ts := lm.GetReadTimestamp(srv.Clock().Now())
+		state := lm.findDescriptorState(descpb.ID(id), false)
+		require.NotNilf(t, state, "the descriptor was not leased yet")
+		ld, _, err := state.findForTimestamp(ctx, ts)
+		require.NoError(t, err)
+		heldDescriptors = append(heldDescriptors, ld)
+		return ld.GetVersion()
+	}
 
-	//lm := srv.LeaseManager().(*Manager)
-	blockUpdates.Store(true)
+	waitForTimestampChange := func() {
+		initial := lm.GetSafeReplicationTS()
+		rangeFeedCh <- struct{}{}
+		testutils.SucceedsSoon(t, func() error {
+			if lm.GetSafeReplicationTS() == initial {
+				return errors.New("timestamp did not change")
+			}
+			return nil
+		})
+	}
 	// Allow one descriptor version to be published and a range feed check point.
 	updateCh <- struct{}{}
-	rangeFeedCh <- struct{}{}
-	// Note: The schema_changes internal table does not use leased descriptors,
-	// but we can force it by intentionally caching t1 as a leased descriptor.
-	schemaChangeDesc := r.QueryStr(t, "SELECT name, type, state FROM t1, crdb_internal.schema_changes")
-	require.Equal(t, schemaChangeDesc, [][]string{{`t1`, `COLUMN`, `WRITE_ONLY`}})
+	waitForTimestampChange()
+	initialVersion := getDescriptorVersion()
 	// A new version is published, but it won't be visible to our transaction yet.
 	updateCh <- struct{}{}
-	schemaChangeDesc = r.QueryStr(t, "SELECT name, type, state FROM t1, crdb_internal.schema_changes")
-	require.Equal(t, schemaChangeDesc, [][]string{{`t1`, `COLUMN`, `WRITE_ONLY`}})
-	rangeFeedCh <- struct{}{}
+	nextVersion := getDescriptorVersion()
+	require.Equalf(t, initialVersion, nextVersion, "new version should not be leased yet")
+	waitForTimestampChange()
 	// A new version should show up after the range feed check point.
-	schemaChangeDesc = r.QueryStr(t, "SELECT name, type, state FROM t1, crdb_internal.schema_changes")
-	require.Len(t, schemaChangeDesc, 0)
+	nextVersion = getDescriptorVersion()
+	require.Equalf(t, initialVersion+1, nextVersion, "new version should be leasable now")
+	releaseHeldDescriptors()
 	close(updateCh)
 	close(rangeFeedCh)
 	require.NoError(t, grp.Wait())

--- a/pkg/sql/catalog/lease/testutils.go
+++ b/pkg/sql/catalog/lease/testutils.go
@@ -68,9 +68,9 @@ type ManagerTestingKnobs struct {
 	// has been reset.
 	RangeFeedResetChannel chan struct{}
 
-	// TestingOnRangeFeedCheckPoint is invoked when a range feed checkpoint is
+	// TestingOnUpdateReadTimestamp is invoked when a range feed checkpoint is
 	// hit.
-	TestingOnRangeFeedCheckPoint func()
+	TestingOnUpdateReadTimestamp func(timestamp hlc.Timestamp)
 
 	LeaseStoreTestingKnobs StorageTestingKnobs
 }

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -2418,6 +2418,10 @@ func (ex *connExecutor) commitSQLTransactionInternal(ctx context.Context) (retEr
 		}
 	}
 
+	if err := ex.extraTxnState.descCollection.EmitDescriptorUpdatesKey(ctx, ex.state.mu.txn); err != nil {
+		return err
+	}
+
 	if err := ex.state.mu.txn.Commit(ctx); err != nil {
 		return err
 	}

--- a/pkg/sql/internal.go
+++ b/pkg/sql/internal.go
@@ -2049,6 +2049,9 @@ func (ief *InternalDB) txn(
 			if err != nil {
 				return err
 			}
+			if err := descsCol.EmitDescriptorUpdatesKey(ctx, kvTxn); err != nil {
+				return err
+			}
 			// We check this testing condition here since a retry cannot be generated
 			// after a successful commit. Since we commit below, this is our last
 			// chance to generate a retry for users of (*InternalDB).Txn.

--- a/pkg/sql/vecindex/manager_test.go
+++ b/pkg/sql/vecindex/manager_test.go
@@ -117,11 +117,6 @@ func TestVectorManager(t *testing.T) {
 	}).BuildCreatedMutable()
 
 	err := internalDB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) (err error) {
-		defer func() {
-			if err == nil {
-				err = txn.KV().Commit(ctx)
-			}
-		}()
 		b := txn.KV().NewBatch()
 
 		err = txn.Descriptors().WriteDescToBatch(ctx, false, newDB, b)

